File: src/controller/device.cc
// Copyright 2022 The Android Open Source Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "controller/device.h"

#include <string>
#include <string_view>
#include <vector>

#include "model.pb.h"
#include "util/log.h"
#include "util/os_utils.h"
#include "util/string_utils.h"

namespace netsim {
namespace controller {
namespace {

// common_typos_disable
const std::vector<std::string> kDeviceNames{
    "Bear", "Boar", "Buck", "Bull", "Calf", "Cavy", "Colt", "Cony", "Coon",
    "Dauw", "Deer", "Dieb", "Douc", "Dzho", "Euro", "Eyra", "Fawn", "Foal",
    "Gaur", "Gilt", "Goat", "Guib", "Gyal", "Hare", "Hart", "Hind", "Hogg",
    "Ibex", "Joey", "Jomo", "Kine", "Kudu", "Lamb", "Lion", "Maki", "Mara",
    "Mare", "Mico", "Mink", "Moco", "Mohr", "Moke", "Mole", "Mona", "Mule",
    "Musk", "Napu", "Neat", "Nowt", "Oont", "Orca", "Oryx", "Oxen", "Paca",
    "Paco", "Pard", "Peba", "Pika", "Pudu", "Puma", "Quey", "Roan", "Runt",
    "Rusa", "Saki", "Seal", "Skug", "Sore", "Tait", "Tegg", "Titi", "Unau",
    "Urus", "Urva", "Vari", "Vole", "Wolf", "Zati", "Zebu", "Zobo", "Zobu"};

const std::string GetName(std::string_view device_name) {
  unsigned int idx =
      std::hash<std::string_view>()(device_name) % kDeviceNames.size();
  return kDeviceNames.at(idx);
}
}  // namespace

void Device::Patch(const model::Device &request) {
  if (request.has_position()) {
    this->model.mutable_position()->CopyFrom(request.position());
  }
  if (request.has_orientation()) {
    this->model.mutable_orientation()->CopyFrom(request.orientation());
  }
  if (!request.name().empty()) {
    this->model.set_name(request.name());
  }
  for (auto &request_chip_model : request.chips()) {
    auto found = false;
    for (auto &chip : chips) {
      if (chip->KeyComp(request_chip_model)) {
        chip->Patch(request_chip_model);
        found = true;
        break;
      }
    }
    if (!found) {
      BtsLog("Unknown chip in update");
    }
  }
}

bool Device::RemoveChip(common::ChipKind chip_kind,
                        const std::string &chip_id) {
  for (int c = 0; c < model.chips().size(); c++) {
    auto c_kind = model.chips().at(c).chip_kind();
    auto c_id = model.chips().at(c).chip_id();
    if (chip_kind == c_kind && chip_id == c_id) {
      // Entry in chips and model are at same index
      BtsLog("Removing chip kind:%d id:'%s' from %s", c_kind, c_id.c_str(),
             model.name().c_str());
      chips[c]->Remove();
      chips.erase(chips.begin() + c);
      model.mutable_chips()->DeleteSubrange(c, 1);
      return chips.size() == 0;
    }
  }
  BtsLog("Trying to remove unknown chip");
  return chips.size() == 0;
}

void Device::AddChip(std::shared_ptr<Device> device, std::shared_ptr<Chip> chip,
                     const model::Chip &chip_model) {
  for (auto &c : chips) {
    if (c->KeyComp(chip_model)) {
      BtsLog("Trying to add a duplicate chip, skipping!");
      return;
    }
  }
  chip->Init(std::move(device), chips.size());
  model.mutable_chips()->Add()->CopyFrom(chip_model);
  chips.push_back(std::move(chip));
}
void Device::Reset() {
  this->model.set_visible(true);
  this->model.mutable_position()->Clear();
  this->model.mutable_orientation()->Clear();
  // TODO: Reset chips and radios.
}

std::shared_ptr<Device> CreateDevice(std::string_view name) {
  model::Device model;
  model.set_name(stringutils::AsString(name));
  model.set_visible(true);
  // required sub-messages to simplify ui
  model.mutable_position();
  model.mutable_orientation();
  return std::make_shared<Device>(model);
}

}  // namespace controller
}  // namespace netsim

File: rust/netsim-cxx/src/devices/device.rs
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Device.rs

use frontend_proto::model::State;
use protobuf::Message;

use crate::devices::chip;
use crate::devices::chip::Chip;
use crate::devices::chip::ChipIdentifier;
use crate::devices::chip::FacadeIdentifier;
use frontend_proto::common::ChipKind as ProtoChipKind;
use frontend_proto::model::Device as ProtoDevice;
use frontend_proto::model::Orientation as ProtoOrientation;
use frontend_proto::model::Position as ProtoPosition;
use std::collections::BTreeMap;

pub type DeviceIdentifier = u32;

pub struct Device {
    pub id: DeviceIdentifier,
    pub guid: String,
    pub name: String,
    visible: State,
    pub position: ProtoPosition,
    orientation: ProtoOrientation,
    pub chips: BTreeMap<ChipIdentifier, Chip>,
}
impl Device {
    pub fn new(id: DeviceIdentifier, guid: String, name: String) -> Self {
        Device {
            id,
            guid,
            name,
            visible: State::ON,
            position: ProtoPosition::new(),
            orientation: ProtoOrientation::new(),
            chips: BTreeMap::new(),
        }
    }
}

#[derive(Debug)]
pub struct AddChipResult {
    pub device_id: DeviceIdentifier,
    pub chip_id: ChipIdentifier,
    pub facade_id: FacadeIdentifier,
}

impl Device {
    pub fn get(&self) -> Result<ProtoDevice, String> {
        let mut device = ProtoDevice::new();
        device.id = self.id;
        device.name = self.name.clone();
        device.visible = self.visible.into();
        device.position = protobuf::MessageField::from(Some(self.position.clone()));
        device.orientation = protobuf::MessageField::from(Some(self.orientation.clone()));
        for chip in self.chips.values() {
            device.chips.push(chip.get()?);
        }
        Ok(device)
    }

    /// Patch a device and its chips.
    pub fn patch(&mut self, patch: &ProtoDevice) -> Result<(), String> {
        let patch_visible = patch.visible.enum_value_or_default();
        if patch_visible != State::UNKNOWN {
            self.visible = patch_visible;
        }
        if patch.position.is_some() {
            self.position.clone_from(&patch.position);
        }
        if patch.orientation.is_some() {
            self.orientation.clone_from(&patch.orientation);
        }
        // iterate over patched ProtoChip entries and patch matching chip
        for patch_chip in patch.chips.iter() {
            let mut patch_chip_kind = patch_chip.kind.enum_value_or_default();
            // Check if chip is given when kind is not given.
            // TODO: Fix patch device request body in CLI to include ChipKind, and remove if block below.
            if patch_chip_kind == ProtoChipKind::UNSPECIFIED {
                if patch_chip.has_bt() {
                    patch_chip_kind = ProtoChipKind::BLUETOOTH;
                } else if patch_chip.has_wifi() {
                    patch_chip_kind = ProtoChipKind::WIFI;
                } else if patch_chip.has_uwb() {
                    patch_chip_kind = ProtoChipKind::UWB;
                } else {
                    break;
                }
            }
            let patch_chip_name = &patch_chip.name;
            // Find the matching chip and patch the proto chip
            for chip in self.chips.values_mut() {
                if (patch_chip_name.is_empty() || chip.name.eq(patch_chip_name))
                    && chip.kind == patch_chip_kind
                {
                    chip.patch(patch_chip)?;
                    break; // next proto chip
                }
            }
        }
        Ok(())
    }

    /// Remove a chip from a device.
    pub fn remove_chip(
        &mut self,
        chip_id: ChipIdentifier,
    ) -> Result<(Option<FacadeIdentifier>, ProtoChipKind), String> {
        let (facade_id, kind) = {
            if let Some(chip) = self.chips.get_mut(&chip_id) {
                (chip.facade_id, chip.kind)
            } else {
                return Err(format!("RemoveChip chip id {chip_id} not found"));
            }
        };
        match self.chips.remove(&chip_id) {
            Some(_) => Ok((facade_id, kind)),
            None => Err(format!("Key {chip_id} not found in Hashmap")),
        }
    }

    pub fn add_chip(
        &mut self,
        device_name: &str,
        chip_kind: ProtoChipKind,
        chip_name: &str,
        chip_manufacturer: &str,
        chip_product_name: &str,
    ) -> Result<(DeviceIdentifier, ChipIdentifier), String> {
        for chip in self.chips.values() {
            if chip.kind == chip_kind && chip.name == chip_name {
                return Err(format!("Device::AddChip - duplicate at id {}, skipping.", chip.id));
            }
        }
        let chip = chip::chip_new(
            chip_kind,
            chip_name,
            device_name,
            chip_manufacturer,
            chip_product_name,
        )?;
        let chip_id = chip.id;
        self.chips.insert(chip_id, chip);
        Ok((self.id, chip_id))
    }

    /// Reset a device to its default state.
    pub fn reset(&mut self) -> Result<(), String> {
        self.visible = State::ON;
        self.position.clear();
        self.orientation.clear();
        for chip in self.chips.values_mut() {
            chip.reset()?;
        }
        Ok(())
    }
}
